sqlitepool/sqlitepool.go

```diff
@@ -0,0 +1,318 @@
+// Package sqlitepool implements a pool of SQLite database connections.
+package sqlitepool
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"strings"
+
+	"github.com/tailscale/sqlite/cgosqlite"
+	"github.com/tailscale/sqlite/sqliteh"
+)
+
+// A Pool is a fixed-size pool of SQLite database connections.
+// One is reserved for writable transactions, the others are
+// used for read-only transactions.
+type Pool struct {
+	poolSize    int
+	rwConnFree  chan *conn // cap == 1
+	roConnsFree chan *conn // cap == poolSize-1
+	tracer      sqliteh.Tracer
+	closed      chan struct{}
+}
+
+type conn struct {
+	pool  *Pool
+	db    sqliteh.DB
+	stmts map[string]sqliteh.Stmt // persistent statements on db
+	id    sqliteh.TraceConnID
+}
+
+// NewPool creates a Pool of poolSize database connections.
+//
+// For each connection, initFn is called to initialize the connection.
+// Tracer is used to report statistics about the use of the Pool.
+func NewPool(filename string, poolSize int, initFn func(sqliteh.DB) error, tracer sqliteh.Tracer) (p *Pool, err error) {
+	defer func() {
+		if err != nil {
+			err = fmt.Errorf("sqlitepool.NewPool: %w", err)
+			select {
+			case conn := <-p.rwConnFree:
+				conn.db.Close()
+			default:
+			}
+			for conn := range p.roConnsFree {
+				conn.db.Close()
+			}
+		}
+	}()
+	if poolSize < 2 {
+		return nil, fmt.Errorf("poolSize=%d is too small", poolSize)
+	}
+	p = &Pool{
+		poolSize:    poolSize,
+		rwConnFree:  make(chan *conn, 1),
+		roConnsFree: make(chan *conn, poolSize-1),
+		tracer:      tracer,
+		closed:      make(chan struct{}),
+	}
+	for i := 0; i < poolSize; i++ {
+		db, err := cgosqlite.Open(filename, sqliteh.OpenFlagsDefault, "")
```

Comment:
Seems like there's no way to pass open flags or VFS using NewPool?

Reply:
not yet. may add in the future.